qcore/siteamp_models.py

```diff
@@ -58,107 +52,553 @@ def amplification_uncertainty(
+    # fs1100 - fs_vpga for T=0
+    fs_high_0 = _compute_fs_value(0, 1100.0, pga, c10, k1, k2)  # fs_high for T=0
+    fs_vpga_0 = _compute_fs_value(0, vpga, pga, c10, k1, k2)  # fs_auto for T=0
+    a1100 = pga * np.exp(fs_high_0 - fs_vpga_0)
+
+    # Calculate amplification factors for each period
+    ampf0 = np.zeros(freqs.size, dtype=np.float64)
+    t_idx = 0
+    while freqs[t_idx] > flowcap and t_idx < freqs.size:
+        fs_site = _compute_fs_value(t_idx, vsite, a1100, c10, k1, k2)
+        fs_ref = _compute_fs_value(t_idx, vref, a1100, c10, k1, k2)
+        ampf0[t_idx] = np.exp(fs_site - fs_ref)
+        t_idx += 1
+    ampf0[t_idx:] = ampf0[t_idx]
+
+    return ampf0
+
+
+class CBModelVersion(Enum):
+    """Campbell and Bozorgnia model versions"""
+
+    CB2008 = 2008
+    CB2014 = 2014
+
+
+@njit(cache=True, parallel=True)
+def _cb_amp_multi(
+    vref: np.ndarray,
+    vsite: np.ndarray,
+    vpga: np.ndarray,
+    pga: np.ndarray,
+    version: int,
+    flowcap: float,
+    freqs: np.ndarray,
+) -> np.ndarray:
+    """
+    Numba version of cb_amp that processes multiple parameter sets.
+
+    Parameters
+    ----------
+    vref : array_like
+        Reference Vs30 values (m/s) - shape (N,)
+    vsite : array_like
+        Site Vs30 values (m/s) - shape (N,)
+    vpga : array_like
+        Vs30 values for PGA calculation (m/s) - shape (N,)
+    pga : array_like
+        Peak ground acceleration values (g) - shape (N,)
+    version : CBModelVersion
+        CB version (2008 or 2014)
+    flowcap : float
+        Flow capacity constraint
+    freqs : np.ndarray
+        Frequencies to compute amplification values for using model
+        explicitly.
+
+    Returns
+    -------
+    np.ndarray
+        Amplification factors, shape (N, output_length)
+        where N is the number of input parameter sets
+        and output_length depends on dt and n
+    """
 
-    # calculate factor for each period
-    it = (
-        exp(fs_auto(T, vsite)(T, vsite, a1100) - fs_auto(T, vref)(T, vref, a1100))
-        for T in range(freqs.size)
-    )
-    ampf0 = np.fromiter(it, dtype=np.float64)
-    try:
-        # T is the first occurance of a value <= flowcap
-        T = np.flatnonzero((freqs <= flowcap))[0]
-        ampf0[T:] = ampf0[T]
-    except IndexError:
-        pass
+    # Convert inputs to arrays and get dimensions
+    vref_arr = np.asarray(vref, dtype=np.float64)
+    vsite_arr = np.asarray(vsite, dtype=np.float64)
+    vpga_arr = np.asarray(vpga, dtype=np.float64)
+    pga_arr = np.asarray(pga, dtype=np.float64)
+
+    # Check that all arrays have the same shape
+    n_cases = vref_arr.size
+
+    # Flatten arrays to handle both 1D and scalar inputs
+    vref_flat = vref_arr.flatten()
+    vsite_flat = vsite_arr.flatten()
+    vpga_flat = vpga_arr.flatten()
+    pga_flat = pga_arr.flatten()
+
+    # Determine output size by running one case
+    # This is needed since we don't know the output size a priori
```

Comment:
I think these can be removed now as the n_cases seems to come from the vref_arr.size now